src/Entities/Player.cpp

```diff
@@ -2643,6 +2643,93 @@ void cPlayer::SendBlocksAround(int a_BlockX, int a_BlockY, int a_BlockZ, int a_R
 
 bool cPlayer::PlaceBlocks(const sSetBlockVector & a_Blocks)
 {
+	// Check to see if any entity intersects any block being placed
+	class DoesIntersectBlock : public cEntityCallback
+	{
+	public:
+		const sSetBlockVector & m_Blocks;
+		cWorld * m_World;
+
+		// The distance inside the block the entity can still be.
+		const double EPSILON = 0.05;
+
+		DoesIntersectBlock(const sSetBlockVector & a_Blocks, cWorld * a_World) :
+			m_Blocks(a_Blocks)
+			, m_World(a_World)
+		{
+		}
+
+		virtual bool Item(cEntity * a_Entity) override
+		{
+			cBoundingBox EntBox(a_Entity->GetPosition(), a_Entity->GetWidth() / 2, a_Entity->GetHeight());
+			for (auto blk: m_Blocks)
+			{
+				cBlockHandler * BlockHandler = cBlockInfo::GetHandler(blk.m_BlockType);
+				int x = blk.GetX();
+				int y = blk.GetY();
+				int z = blk.GetZ();
+				cBoundingBox BlockBox = BlockHandler->GetPlacementCollisionBox(
+					m_World->GetBlock(x - 1, y, z),
+					m_World->GetBlock(x + 1, y, z),
+					m_World->GetBlock(x, y - 1, z),
+					m_World->GetBlock(x, y + 1, z),
+					m_World->GetBlock(x, y, z - 1),
+					m_World->GetBlock(x, y, z + 1)
+				);
+				BlockBox.Move(x, y, z);
+
+				// Put in a little bit of wiggle room
+				BlockBox.Expand(-EPSILON, -EPSILON, -EPSILON);
+				if (EntBox.DoesIntersect(BlockBox))
+				{
+					return true;
+				}
+			}
+			return false;
+		}
+	} Callback(a_Blocks, m_World);
+
+	// Compute the bounding box for each block to be placed
+	cBoundingBox PlacingBounds(0, 0, 0, 0, 0, 0);
+	bool HasInitializedBounds = false;
+	for (auto blk: a_Blocks)
+	{
+		cBlockHandler * BlockHandler = cBlockInfo::GetHandler(blk.m_BlockType);
+		int x = blk.GetX();
+		int y = blk.GetY();
+		int z = blk.GetZ();
+		cBoundingBox BlockBox = BlockHandler->GetPlacementCollisionBox(
+			m_World->GetBlock(x - 1, y, z),
+			m_World->GetBlock(x + 1, y, z),
+			m_World->GetBlock(x, y - 1, z),
+			m_World->GetBlock(x, y + 1, z),
+			m_World->GetBlock(x, y, z - 1),
+			m_World->GetBlock(x, y, z + 1)
+		);
+		BlockBox.Move(x, y, z);
+
+		if (HasInitializedBounds)
+		{
+			PlacingBounds = PlacingBounds.Union(BlockBox);
+		}
+		else
+		{
+			PlacingBounds = BlockBox;
+			HasInitializedBounds = true;
+		}
+	}
+
+	// See if any entities in that bounding box collide with anyone
+	if (!m_World->ForEachEntityInBox(PlacingBounds, Callback))
+	{
+		// Abort - re-send all the current blocks in the a_Blocks' coords to the client:
+		for (auto blk2: a_Blocks)
+		{
+			m_World->SendBlockTo(blk2.GetX(), blk2.GetY(), blk2.GetZ(), this);
+		}
+		return false;
+	}
+
```

Comment:
This whole block is already rather large, it would be better to move it into a separate function.